### ISSUE TITLE
Client WS suddenly stopped working with error "VH default: iface \ufffdA\ufffd port 0 DOESN'T EXIST"

### ISSUE THREAD

**Issue by chinhodado** [warmcat/libwebsockets#2361]
I have a WS client that was working fine until a few days ago, where without any code change it would fail with this error:

```
[2021/07/23 14:52:08:1923] I: Initial logging level 2047
[2021/07/23 14:52:08:1925] I: Libwebsockets version: 4.0.1 unknown-build-hash
[2021/07/23 14:52:08:1926] I: IPV6 compiled in and enabled
[2021/07/23 14:52:08:1927] I:  LWS_DEF_HEADER_LEN    : 4096
[2021/07/23 14:52:08:1927] I:  LWS_MAX_SMP           : 1
[2021/07/23 14:52:08:1927] I:  sizeof (*info)        : 680
[2021/07/23 14:52:08:1928] I:  SYSTEM_RANDOM_FILEPATH: '/dev/urandom'
[2021/07/23 14:52:08:1929] I:  HTTP2 support         : available
[2021/07/23 14:52:08:1930] D: _realloc: size 9336: context
[2021/07/23 14:52:08:1930] I: context created
[2021/07/23 14:52:08:1931] I: Using event loop: poll
[2021/07/23 14:52:08:1931] I: Default ALPN advertisment: h2,http/1.1
[2021/07/23 14:52:08:1932] I:  default timeout (secs): 20
[2021/07/23 14:52:08:1932] I:  Threads: 1 each 1024 fds
[2021/07/23 14:52:08:1933] I:  mem: context:          8160 B (4064 ctx + (1 thr x 4096))
[2021/07/23 14:52:08:1934] I:  mem: http hdr size:   (4096 + 976), max count 1024
[2021/07/23 14:52:08:1934] D: _realloc: size 8192: fds table
[2021/07/23 14:52:08:1935] I:  mem: pollfd map:       8192 B
[2021/07/23 14:52:08:1938] D: _realloc: size 8192: lws_lookup
[2021/07/23 14:52:08:1938] I:  mem: platform fd map:  8192 B
[2021/07/23 14:52:08:1939] D: _realloc: size 1176: event pipe wsi
[2021/07/23 14:52:08:1939] D: lws_role_transition: 0x7f3bc80072c0: wsistate 0x200, ops pipe
[2021/07/23 14:52:08:1940] D: event pipe fd 20
[2021/07/23 14:52:08:1940] D: __insert_wsi_socket_into_fds: 0x7f3bc80072c0: tsi=0, sock=20, pos-in-fds=0
[2021/07/23 14:52:08:1941] I:  Compiled with OpenSSL support
[2021/07/23 14:52:08:1942] I:  SSL disabled: no LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT
[2021/07/23 14:52:08:1942] I:  canonical_hostname = mymachine
[2021/07/23 14:52:08:1943] D: _realloc: size 1096: create vhost
[2021/07/23 14:52:08:1943] D: _realloc: size 168: vhost-specific plugin table
[2021/07/23 14:52:08:1944] D: _realloc: size 48: same vh list
[2021/07/23 14:52:08:1945] I: Creating Vhost 'default' port 0, 2 protocols, IPv6 on
[2021/07/23 14:52:08:1953] D:  interface lo vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:52:08:1954] D:  interface eth0 vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:52:08:1956] D:  interface lo vs \ufffdA\ufffd (fam 2) ipv6 1
[2021/07/23 14:52:08:1957] D:  interface eth0 vs \ufffdA\ufffd (fam 2) ipv6 1
[2021/07/23 14:52:08:1957] D:  interface lo vs \ufffdA\ufffd (fam 10) ipv6 1
[2021/07/23 14:52:08:1959] D:  interface eth0 vs \ufffdA\ufffd (fam 10) ipv6 1
[2021/07/23 14:52:08:1959] I: lws_socket_bind: netif \ufffdA\ufffd: Does not exist
[2021/07/23 14:52:08:1960] D: initial if check says -1
[2021/07/23 14:52:08:1960] D: deferred iface: adding vh default
[2021/07/23 14:52:08:1962] E: VH default: iface \ufffdA\ufffd port 0 DOESN'T EXIST
[2021/07/23 14:52:08:1963] I:  mem: per-conn:         1176 bytes + protocol rx buf
[2021/07/23 14:52:08:1963] I: lws_plat_drop_app_privileges: not changing group
[2021/07/23 14:52:08:1964] I: lws_plat_drop_app_privileges: not changing user
[2021/07/23 14:52:08:1965] I: lws_cancel_service
[2021/07/23 14:52:08:1966] D: _lws_state_transition: system: changed 1 'CONTEXT_CREATED' -> 2 'INITIALIZED'
[2021/07/23 14:52:08:1966] D: _lws_state_transition: system: changed 2 'INITIALIZED' -> 3 'IFACE_COLDPLUG'
[2021/07/23 14:52:08:1967] D: _lws_state_transition: system: changed 3 'IFACE_COLDPLUG' -> 4 'DHCP'
[2021/07/23 14:52:08:1968] D: _lws_state_transition: system: changed 4 'DHCP' -> 5 'TIME_VALID'
[2021/07/23 14:52:08:1968] I: lws_state_notify_protocol_init: doing protocol init on POLICY_VALID
[2021/07/23 14:52:08:1970] I: lws_protocol_init
[2021/07/23 14:52:08:1971] D: _lws_state_transition: system: changed 5 'TIME_VALID' -> 6 'POLICY_VALID'
[2021/07/23 14:52:08:1971] D: _lws_state_transition: system: changed 6 'POLICY_VALID' -> 7 'REGISTERED'
[2021/07/23 14:52:08:1972] D: _lws_state_transition: system: changed 7 'REGISTERED' -> 8 'AUTH1'
[2021/07/23 14:52:08:1973] D: _lws_state_transition: system: changed 8 'AUTH1' -> 9 'AUTH2'
[2021/07/23 14:52:08:1975] D: _lws_state_transition: system: changed 9 'AUTH2' -> 10 'OPERATIONAL'
[2021/07/23 14:52:08:1976] I: lws_state_transition_steps: CONTEXT_CREATED -> OPERATIONAL
[2021/07/23 14:52:38:1978] D: deferred iface: checking if on vh default
[2021/07/23 14:52:38:1980] D:  interface lo vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:52:38:1980] D:  interface eth0 vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:52:38:1980] D:  interface lo vs \ufffdA\ufffd (fam 2) ipv6 1
[2021/07/23 14:52:38:1980] D:  interface eth0 vs \ufffdA\ufffd (fam 2) ipv6 1
[2021/07/23 14:52:38:1980] D:  interface lo vs \ufffdA\ufffd (fam 10) ipv6 1
[2021/07/23 14:52:38:1981] D:  interface eth0 vs \ufffdA\ufffd (fam 10) ipv6 1
[2021/07/23 14:52:38:1981] I: lws_socket_bind: netif \ufffdA\ufffd: Does not exist
[2021/07/23 14:52:38:1981] D: initial if check says -1
[2021/07/23 14:53:08:2012] D: deferred iface: checking if on vh default
[2021/07/23 14:53:08:2016] D:  interface lo vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:53:08:2016] D:  interface eth0 vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:53:08:2016] D:  interface lo vs \ufffdA\ufffd (fam 2) ipv6 1
[2021/07/23 14:53:08:2016] D:  interface eth0 vs \ufffdA\ufffd (fam 2) ipv6 1
[2021/07/23 14:53:08:2016] D:  interface lo vs \ufffdA\ufffd (fam 10) ipv6 1
[2021/07/23 14:53:08:2016] D:  interface eth0 vs \ufffdA\ufffd (fam 10) ipv6 1
[2021/07/23 14:53:08:2016] I: lws_socket_bind: netif \ufffdA\ufffd: Does not exist
[2021/07/23 14:53:08:2016] D: initial if check says -1
[2021/07/23 14:53:38:2050] D: deferred iface: checking if on vh default
[2021/07/23 14:53:38:2052] D:  interface lo vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:53:38:2052] D:  interface eth0 vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:53:38:2052] D:  interface lo vs \ufffdA\ufffd (fam 2) ipv6 1
[2021/07/23 14:53:38:2052] D:  interface eth0 vs \ufffdA\ufffd (fam 2) ipv6 1
[2021/07/23 14:53:38:2052] D:  interface lo vs \ufffdA\ufffd (fam 10) ipv6 1
[2021/07/23 14:53:38:2052] D:  interface eth0 vs \ufffdA\ufffd (fam 10) ipv6 1
[2021/07/23 14:53:38:2052] I: lws_socket_bind: netif \ufffdA\ufffd: Does not exist
[2021/07/23 14:53:38:2052] D: initial if check says -1
```

What puzzles me is the garbage text for the interface name, I don't know where the library got it from.

Here are the machine's interfaces:

```
$ ifconfig
eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.x.x.x  netmask 255.255.252.0  broadcast 10.x.x.x
        inet6 xxxx  prefixlen 64  scopeid 0x20<link>
        ether aa:bb:cc:dd:ee:ff  txqueuelen 1000  (Ethernet)
        RX packets 1784388  bytes 213788406 (203.8 MiB)
        RX errors 0  dropped 4277  overruns 0  frame 0
        TX packets 967272  bytes 1052246500 (1003.5 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 7915002  bytes 527184421 (502.7 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 7915002  bytes 527184421 (502.7 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

and here is the code I used to create the context. I did not specify any manual interface nor vhost:

```
   struct lws_context_creation_info info;
   memset(&info, 0, sizeof info); /* otherwise uninitialized garbage */
   info.port = CONTEXT_PORT_NO_LISTEN; /* we do not run any server */
   info.protocols = ws_protocols;
   info.fd_limit_per_thread = 1 + 1 + 1;
   info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT;
   struct lws_context *context = lws_create_context(&info);
```

How can I figure out what's wrong? I also tried to set an interface manually (by adding `info.iface = "x.x.x.x"` to the above code) but it didn't do anything.

**Comment by lws-team** [comment]
What version of lws is it and what platform?

Leaving .iface as NULL should be OK, it looks like something overwritten

```
[2021/07/23 14:52:08:1945] I: Creating Vhost 'default' port 0, 2 protocols, IPv6 on
[2021/07/23 14:52:08:1953] D:  interface lo vs \ufffdA\ufffd (fam 17) ipv6 1
[2021/07/23 14:52:08:1954] D:  interface eth0 vs \ufffdA\ufffd (fam 17) ipv6 1
```

This fffd character is unicode for an invalid unicode char.

Since it looks like this happens before we return from context creation, it's less likely to be your code

I would

 - check for multiple installs of lws headers, of different versions.  This can happen if you install a distro lws package that puts headers in /usr/include, and later build and install locally that defaults to /usr/local/include.  Then we may end up filling info into the wrong offsets,

 - make clean && make && make install the lws part and then the same to your app part

 - if your code contains threads, check your arrangements for making sure the context has finished creating before you try to use it
 
 - if other pieces to your app, temporarily disable everything except the context creation and see if it changes the situation.

**Comment by chinhodado** [comment]
Indeed, I was using a different versions of libwebsocket. The installed distro package was 4.0.1, and I had checked out the repo at a different version. After I checked out the 4.0.1 tag of the repo and rebuilt, the issue went away. Thank you for the pointers! Not sure why it worked for a long time before running into this error though, since I did not touch the libwebsocket repo or changed the installed package during that time.
